android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/RibCoroutineWorker.kt

```diff
@@ -0,0 +1,187 @@
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.supervisorScope
+import kotlin.coroutines.ContinuationInterceptor
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.coroutineContext
+import kotlin.coroutines.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.intrinsics.intercepted
+import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn
+import kotlin.coroutines.resume
+
+/** A manager or helper class bound to a [CoroutineScope] by using a binder like [bind]. */
+public interface RibCoroutineWorker {
+  /** Called when worker is started. Children coroutines can be launched in [scope]. */
+  @JvmDefault
+  public suspend fun onStart(scope: CoroutineScope) {}
+
+  /** Called when the worker is stopped with the given [cause]. Should be fast, non-blocking and should not throw. */
+  @JvmDefault
+  public fun onStop(cause: Throwable) {}
+}
+
+// ---- Binder ---- //
+
+/**
+ * A handle to interact with [RibCoroutineWorker] binding job. This handle implements [Job], which refers to the
+ * completion of [RibCoroutineWorker.onStart]. It can be [joined][join] to make sure `onStart` finished. Note that
+ * children coroutines launched in the [CoroutineScope] passed on to `onStart` are not waited: worker is considered
+ * bound when `onStart` finishes.
+ */
+public sealed interface BindWorkerHandle : Job {
+  /** Unbinds the worker. */
+  public fun unbind(): Job
+}
+
+/**
+ * Binds [worker] in a scope that is a child of the [CoroutineScope] receiver.
+ *
+ * The binding operation runs [RibCoroutineWorker.onStart] in a context inherited from the [CoroutineScope] receiver, but
+ * with additional [context] elements that is, by default, [RibDispatchers.Default]. This makes the worker run on the
+ * default dispatcher by default. Pass in [EmptyCoroutineContext] instead if you want the worker to not override the
+ * dispatcher in the scope (if any), or pass in a custom dispatcher as [context] to specify a different dispatcher. If
+ * there is no dispatcher in [CoroutineScope] nor in [context], [RibDispatchers.Default] is used.
+ *
+ * The scope passed on to [RibCoroutineWorker.onStart] as a parameter is a child scope of the [CoroutineScope] receiver,
+ * but with the additional [context] elements and a [SupervisorJob].
+ *
+ * Binding a worker is an asynchronous operation. To ensure [RibCoroutineWorker.onStart] is finished, callers can
+ * [join][BindWorkerHandle.join] the resulting [BindWorkerHandle] when in a coroutine:
+ *
+ * ```
+ * val handle = coroutineScope.bind(worker)
+ * handle.join() // wait for onStart to finish
+ * ```
+ */
+@JvmOverloads
+public fun CoroutineScope.bind(
```

Comment:
Btw, what do you think if we add 2 new methods to ease existing binding of [Workers] off main thread (when called on didBecome interactor)

Internally often times interactors will bind a big list a list of workers via plugin factory. This should ease the binding for those instances and easily switch them from UI to RibDispatchers.Default without major changes

Was thinking something like this

public fun CoroutineScope.bindRibWorkers(workers:List<Worker>, context: CoroutineContext = RibDispatchers.Default){
  for(worker in workers){
    this.bindRibWorker(worker, context)
  }
}

public fun CoroutineScope.bindRibWorker(
  worker: Worker,
  context: CoroutineContext = RibDispatchers.Default): BindWorkerHandle{

  val ribCoroutineWorker = worker.asRibCoroutineWorker()
  return bind(ribCoroutineWorker, context)
}